core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartDownscaleGetRecoveryCandidatesHandler.java

```diff
@@ -0,0 +1,187 @@
+
+    @Inject
+    private ClusterApiConnectors clusterApiConnectors;
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private RuntimeVersionService runtimeVersionService;
+
+    @Inject
+    private CloudInstanceIdToInstanceMetaDataConverter cloudInstanceIdToInstanceMetaDataConverter;
+
+    @Inject
+    private InstanceMetaDataToCloudInstanceConverter instanceMetaDataToCloudInstanceConverter;
+
+    @Inject
+    private InstanceMetaDataService instanceMetaDataService;
+
+    @Inject
+    private EventBus eventBus;
+
+    @Override
+    public Class<StopStartDownscaleGetRecoveryCandidatesRequest> type() {
+        return StopStartDownscaleGetRecoveryCandidatesRequest.class;
+    }
+
+    @Override
+    public void accept(Event<StopStartDownscaleGetRecoveryCandidatesRequest> event) {
+        StopStartDownscaleGetRecoveryCandidatesRequest request = event.getData();
+        LOGGER.info("StopStartDownscaleGetRunningInstancesHandler: {}, with request: {}", request.getResourceId(), request);
+
+        if (!request.isFailureRecoveryEnabled()) {
+            StopStartDownscaleGetRecoveryCandidatesResult result = new StopStartDownscaleGetRecoveryCandidatesResult(request.getResourceId(),
+                    Collections.emptyList(), request.getHostGroupName(), request.getHostIds());
+            notify(result, event);
+            return;
+        }
+
+        CloudContext cloudContext = request.getCloudContext();
+
+        try {
+            CloudConnector connector  = cloudPlatformConnectors.get(cloudContext.getPlatformVariant());
+            AuthenticatedContext ac = getAuthenticatedContext(request, cloudContext, connector);
+
+            Stack stack = stackService.getByIdWithLists(request.getResourceId());
+
+            List<InstanceMetadataView> startedInstanceMetadataWithServicesNotRunning = Collections.emptyList();
+
+            List<InstanceMetadataView> startedInstancesMetadata = cloudInstanceIdToInstanceMetaDataConverter.getNotDeletedAndNotZombieInstances(
+                    instanceMetaDataService.getAllAvailableInstanceMetadataViewsByStackId(stack.getId()),
+                    request.getHostGroupName(),
+                    collectStartedInstancesFromCloudProvider(connector, ac, request.getAllInstancesInHostGroup()).stream()
+                            .map(CloudInstance::getInstanceId)
+                            .collect(Collectors.toSet()));
+
+            ClusterHealthService clusterHealthService = clusterApiConnectors.getConnector(stack).clusterHealthService();
+
+            DetailedHostStatuses detailedHostStatuses =
+                    clusterHealthService.getDetailedHostStatuses();
+
+            flowMessageService.fireEventAndLog(stack.getId(), UPDATE_IN_PROGRESS.name(), CLUSTER_SCALING_STOPSTART_DOWNSCALE_IDENTIFYRECOVERYCANDIDATES,
+                    request.getHostGroupName());
+
+            if (clusterHealthService.isClusterManagerRunning()) {
```

Comment:
Should the DetailedHostStatus lookup further up also be inside this check ?

Reply:
Similarly on the upscale side

Reply:
Updated